### PR TITLE
[L1T] phase-2, update trackjet simulation

### PR DESCRIPTION
**Pull request by cecilecaillol**
#### PR description:

Porting local https://github.com/cms-l1t-offline/cmssw/pull/1015

This PR contains all the improvements in the track jet code and the improved tagging of displaced jets (and their selection):
The work was presented in many GTT through the years but a summary is also presented in the last GTT meeting:
https://indico.cern.ch/event/1146596/

Several bug fixes are included and also an optimization of the code itself that makes the code 2 times faster than before. In the last slide there is a validation of the improved code.

### CODE REVIEW

**Comment by cecilecaillol** [comment]
please test

**Comment by cecilecaillol** [comment]
+l1

**Comment by srimanob** [comment]
@cmsbuild please test

Refresh the test again.

File: L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc [hunk 1]
```diff
@@ -200,437 +199,152 @@ void L1TrackJetProducer::produce(Event &iEvent, const EventSetup &iSetup) {
...
-        }
+    for (int i = 0; i < phiBins_; ++i) {
+      for (int j = 0; j < etaBins_; ++j) {
+        epbins_default[i][j].pTtot = 0;
+        epbins_default[i][j].used = false;
+        epbins_default[i][j].numtracks = 0;
+        epbins_default[i][j].numttrks = 0;
+        epbins_default[i][j].numtdtrks = 0;
+        epbins_default[i][j].numttdtrks = 0;
+        epbins_default[i][j].trackidx.clear();
```
Comment (AdrianoDee) [inline L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc:238]: I think default initialization in `struct EtaPhiBin` may spare you this.

**Review by AdrianoDee — COMMENTED** [review]
@cecilecaillol see few comments below:

**Comment by AdrianoDee** [comment]
Also: is there a wf where we may test these updates?

**Comment by srimanob** [comment]
@cms-sw/l1-l2 Do you have responses from @AdrianoDee comment? Thanks.

**Comment by srimanob** [comment]
Hi @cms-sw/l1-l2 
Is this PR needed for 12_5 production?

**Comment by epalencia** [comment]
Hi @srimanob, 
no, it is not strictly needed.

**Comment by AdrianoDee** [comment]
hold
- waiting for comments to be addressed
- not needed for 12_5 production

**Comment by cbotta** [comment]
A comment mostly for for @gkaratha , and @BenjaminRS: we decided that this is not needed for the MC production, cause the L1 sequence for this MC production aims at replacing the simulation with the emulation. You left the simulation code in for trackJets and trackVertex because you need it for developments, but updates to the simulation will not affect what we will run in the production.

**Comment by gkaratha** [comment]
Hi @cbotta . thanks for the notice. In that case we will propagate the improvements to the emulation for later usage.

Best,
george

**Comment by cecilecaillol** [comment]
@gkaratha I have rebased the PR and applied Adriano's comments (except "Is there a reason for this change in the parameter initialization " -> can you please reply?). Can you please check everything makes sense? 

@srimanob @AdrianoDee We would like this PR to be backported to 12_5 for our MC production. Can you please unhold the PR?

**Comment by cecilecaillol** [comment]
please test

**Comment by perrotta** [comment]
> @srimanob @AdrianoDee We would like this PR to be backported to 12_5 for our MC production. Can you please unhold the PR?

@cecilecaillol does this comply with what written by @cbotta in https://github.com/cms-sw/cmssw/pull/38337#issuecomment-1223876844?

**Comment by cecilecaillol** [comment]
> > @srimanob @AdrianoDee We would like this PR to be backported to 12_5 for our MC production. Can you please unhold the PR?
> 
> @cecilecaillol does this comply with what written by @cbotta in [#38337 (comment)](https://github.com/cms-sw/cmssw/pull/38337#issuecomment-1223876844)?

We discussed this with @cbotta yesterday and we would like this PR included in 12_5. We also would like https://github.com/cms-sw/cmssw/pull/38334 to be in (to be reopened and needs some work)

**Comment by perrotta** [comment]
please test

**Comment by gkaratha** [comment]
> @gkaratha I have rebased the PR and applied Adriano's comments (except "Is there a reason for this change in the parameter initialization " -> can you please reply?). Can you please check everything makes sense?
> 
> @srimanob @AdrianoDee We would like this PR to be backported to 12_5 for our MC production. Can you please unhold the PR?

Thanks for adding me in the thread. The tests are failing with header consistency. Is that something we should fix?

File: L1Trigger/L1TTrackMatch/interface/L1TrackJetProducer.h [hunk 2]
```diff
@@ -1,21 +1,16 @@
-#ifndef L1Trigger_L1TTrackMatch_L1TrackJetProducer_HH
-#define L1Trigger_L1TTrackMatch_L1TrackJetProducer_HH
+#pragma once
 #include <cmath>
 #include <cstdlib>
-#include <fstream>
-#include <iostream>
-#include <string>
-#include <vector>
```
Comment (AdrianoDee) [inline L1Trigger/L1TTrackMatch/interface/L1TrackJetProducer.h:8]: @gkaratha the compiler is complaining because you removed `#include <vector>` here and so any of the `std::vector` below is defined.

Comment (AdrianoDee) [reply L1Trigger/L1TTrackMatch/interface/L1TrackJetProducer.h:8]:
```c++

src/L1Trigger/L1TTrackMatch/interface/L1TrackJetProducer.h:1:9: warning: #pragma once in main file
    1 | #pragma once
      |         ^~~~
src/L1Trigger/L1TTrackMatch/interface/L1TrackJetProducer.h:16:8: error: 'vector' in namespace 'std' does not name a template type
   16 |   std::vector<unsigned int> trackidx;
      |        ^~~~~~
src/L1Trigger/L1TTrackMatch/interface/L1TrackJetProducer.h:4:1: note: 'std::vector' is defined in header '<vector>'; did you forget to '#include <vector>'?
    3 | #include <cstdlib>
  +++ |+#include <vector>
```

Comment (gkaratha) [reply L1Trigger/L1TTrackMatch/interface/L1TrackJetProducer.h:8]: Ok should I added back?

Is there away from git to take this PR? (given that is Cecile's)

Comment (AdrianoDee) [reply L1Trigger/L1TTrackMatch/interface/L1TrackJetProducer.h:8]: Yes it should be added back but I think only @cecilecaillol may push the modification. I've added a suggestion above.

File: L1Trigger/L1TTrackMatch/interface/L1TrackJetProducer.h [hunk 3]
```diff
@@ -1,21 +1,16 @@
-#ifndef L1Trigger_L1TTrackMatch_L1TrackJetProducer_HH
-#define L1Trigger_L1TTrackMatch_L1TrackJetProducer_HH
+#pragma once
 #include <cmath>
 #include <cstdlib>
```
Comment (AdrianoDee) [inline L1Trigger/L1TTrackMatch/interface/L1TrackJetProducer.h:3]:
```suggestion
#include <cstdlib>
#include <vector>
```

**Comment by cecilecaillol** [comment]
please test

**Comment by cecilecaillol** [comment]
please test

**Comment by cecilecaillol** [comment]
+l1

**Comment by cecilecaillol** [comment]
@AdrianoDee Can we unhold this PR? We have addressed the comments and would like it to be backported to 12_5 for our MC production.

**Comment by AdrianoDee** [comment]
unhold
(review/developments restarted)

**Comment by AdrianoDee** [comment]
+upgrade
- tests run, no difference shown;
- minor comment addressed;
- clean-up and restructuring improving timing performance of a factor 2, as documented (among others) [here](https://indico.cern.ch/event/1146596/contributions/4843520/attachments/2430103/4168746/trackjet_summary.pdf).

File: L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc [hunk 4]
```diff
@@ -200,437 +201,140 @@ void L1TrackJetProducer::produce(Event &iEvent, const EventSetup &iSetup) {
...
     MaxZBin mzb;
+    mzb.znum = -1;
+    mzb.nclust = -1;
+    mzb.ht = -1;
+    EtaPhiBin epbins_default[phiBins_][etaBins_];  // create grid of phiBins
+
+    float phi = -1.0 * M_PI;
+    float eta = -1.0 * trkEtaMax_;
+    for (int i = 0; i < phiBins_; ++i) {
+      eta = -1.0 * trkEtaMax_;
```
Comment (perrotta) [inline L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc:212]:
```suggestion
    for (int i = 0; i < phiBins_; ++i) {
      float eta = -1.0 * trkEtaMax_;
```

File: L1Trigger/L1TTrackMatch/interface/L1TrackJetProducer.h [hunk 5]
```diff
@@ -1,21 +1,18 @@
-#ifndef L1Trigger_L1TTrackMatch_L1TrackJetProducer_HH
```
Comment (perrotta) [inline L1Trigger/L1TTrackMatch/interface/L1TrackJetProducer.h:1]: Why did you remove the header guard and replaced it with `#pragma once`?
I'm asking because it is not the usual way to deal with it in our code...

File: L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc [hunk 6]
```diff
@@ -18,6 +19,7 @@
 #include "FWCore/Framework/interface/Frameworkfwd.h"
 #include "FWCore/Framework/interface/stream/EDProducer.h"
 #include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/ESHandle.h"
```
Comment (perrotta) [inline L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc:22]: Where it is needed?

File: L1Trigger/L1TTrackMatch/interface/L1TrackJetProducer.h (hunk 5 above)
Comment (perrotta) [reply L1Trigger/L1TTrackMatch/interface/L1TrackJetProducer.h:1]: In any case, this file only contains definitions for structs and a few inline functions, and it does not define the L1TrackJetProducer (correctly defined in L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc instead).
As of now, this file is only included by L1TrackJetProducer.cc: if it has to remain as such, it could more conveniently be inserted therein. As far as I understand you are expecting some wider usage for the struct and inline functions defined here: if so, please rename the file in order to make it clear that this is not the header file with the interface for the L1TrackJetProducer

File: L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc [hunk 7]
```diff
@@ -200,437 +201,140 @@ void L1TrackJetProducer::produce(Event &iEvent, const EventSetup &iSetup) {
...
+      L2clusters.clear();
+
+      // fill grid
+      for (unsigned int k = 0; k < L1TrkPtrs_.size(); ++k) {
+        float trkpt = L1TrkPtrs_[k]->momentum().perp();
+        float trketa = L1TrkPtrs_[k]->momentum().eta();
+        float trkphi = L1TrkPtrs_[k]->momentum().phi();
+        float trkZ = L1TrkPtrs_[k]->z0();
+        if (zmax < trkZ)
           continue;
```
Comment (perrotta) [inline L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc:250]: Please don't compute `trkpt` `trketa` `trkphi` if the conditions on thkZ make them useless:
```suggestion
        float trkZ = L1TrkPtrs_[k]->z0();
        if (zmax < trkZ)
          continue;
        if (zmin > trkZ)
          continue;
        if (zbin==0 && zmin == trkZ)
          continue;
        float trkpt = L1TrkPtrs_[k]->momentum().perp();
        float trketa = L1TrkPtrs_[k]->momentum().eta();
        float trkphi = L1TrkPtrs_[k]->momentum().phi();
```

File: L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc [hunk 8]
```diff
@@ -200,437 +201,140 @@ void L1TrackJetProducer::produce(Event &iEvent, const EventSetup &iSetup) {
...
-
-        TkJet trkJet(jetP4, L1TrackAssocJet, mzb.zbincenter, mzb.clusters[j].numtracks, 0, totalDisptrk, 0, isDispJet);
-        if (!L1TrackAssocJet.empty()) {
-          L1L1TrackJetProducer->push_back(trkJet);
+        if (zbin == 0) {
+          if (zmin > trkZ)
+            continue;
+        } else {
+          if (zmin >= trkZ)
+            continue;
```
Comment (perrotta) [inline L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc:256]:
```suggestion
```

File: L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc [hunk 9]
```diff
@@ -200,437 +201,140 @@ void L1TrackJetProducer::produce(Event &iEvent, const EventSetup &iSetup) {
...
+      float jetPhi = mzb.clusters[j].phi;
+      float jetPt = mzb.clusters[j].pTtot;
+      float jetPx = jetPt * cos(jetPhi);
+      float jetPy = jetPt * sin(jetPhi);
+      float jetPz = jetPt * sinh(jetEta);
+      float jetP = jetPt * cosh(jetEta);
+      int totalDisptrk = mzb.clusters[j].numtdtrks;
+      bool isDispJet = false;
+      if (totalDisptrk > nDisplacedTracks_ || totalDisptrk == nDisplacedTracks_)
+        isDispJet = true;
```
Comment (perrotta) [inline L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc:320]:
```suggestion
      bool isDispJet = (totalDisptrk > nDisplacedTracks_ || totalDisptrk == nDisplacedTracks_);
```

File: L1Trigger/L1TTrackMatch/python/l1tTrackJets_cfi.py [hunk 10]
```diff
@@ -3,57 +3,74 @@
...
-	nStubs4Displacedbend=cms.double(1.7),
+	nStubs4DisplacedBend=cms.double(1.7),
 	nStubs5DisplacedChi2=cms.double(2.75),
-	nStubs5Displacedbend=cms.double(3.5),
-	nDisplacedTracks=cms.int32(2) #Number of displaced tracks required per jet
+	nStubs5DisplacedBend=cms.double(3.5),
+        nDisplacedTracks=cms.int32(2)
 )
 
 l1tTrackJetsExtended = cms.EDProducer('L1TrackJetProducer',
```
Comment (perrotta) [inline L1Trigger/L1TTrackMatch/python/l1tTrackJets_cfi.py:34]: This should be better cloned from the previous `l1tTrackJets`, so that common parameters do not have to be manually copied every time a modification is needed, thus reducing the possibility of mistakes.

**Review by perrotta — COMMENTED** [review]
Besides what listed below I would suggest to fully implement the `fillDescriptions` method, already defined in https://github.com/cms-sw/cmssw/blob/2854e1c70cd9100809a06737ff092f6c06fac797/L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc#L364

**Comment by cecilecaillol** [comment]
please test

File: L1Trigger/L1TTrackMatch/python/l1tTrackJets_cfi.py (hunk 10 above)
Comment (cecilecaillol) [reply L1Trigger/L1TTrackMatch/python/l1tTrackJets_cfi.py]: Done

File: L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc (hunk 9 above)
Comment (cecilecaillol) [reply L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc]: Done

File: L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc (hunk 8 above)
Comment (cecilecaillol) [reply L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc]: Done

File: L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc (hunk 7 above)
Comment (cecilecaillol) [reply L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc]: Done

File: L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc (hunk 6 above)
Comment (cecilecaillol) [reply L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc]: Removed

File: L1Trigger/L1TTrackMatch/interface/L1TrackJetProducer.h (hunk 5 above)
Comment (cecilecaillol) [reply L1Trigger/L1TTrackMatch/interface/L1TrackJetProducer.h:1]: Done

File: L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc (hunk 4 above)
Comment (cecilecaillol) [reply L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc]: Done

**Comment by perrotta** [comment]
@cms-sw/upgrade-l2 you already signed this once, is the PR still fine with you after the latest updates?

**Comment by AdrianoDee** [comment]
+upgrade

**Comment by perrotta** [comment]
please test with #39397
(to refresh the 10 days old tests and run them together with the lastly merged L1T Phase2 PR)

**Comment by perrotta** [comment]
+1
